### PR TITLE
Cross-Device Sync, AS 3.0 Update & Event Send Times

### PR DESCRIPTION
**Pull request by sebastian24**

### CODE REVIEW

**Comment by sebastian24** [comment]
https://media.giphy.com/media/3o6ZsXE8RGrhnYRrDG/giphy.gif

File: spilsdk/src/main/java/com/spilgames/spilsdk/userdata/UserDataManager.java [hunk 1]
```diff
@@ -0,0 +1,805 @@
...
+            e.printStackTrace();
+        }
+
+        //Processing Remote User Data
+        try {
+            //Remote PlayerData
+            JSONObject remotePlayerData = new JSONObject();
+
+            for (PlayerCurrency receivedPlayerCurrency : receivedWallet.getCurrenciesMap().values()) {
+                for (Currency currency : spilGameData.getCurrenciesMap().values()) {
```
Comment (svenvanhoenderdal) [inline spilsdk/src/main/java/com/spilgames/spilsdk/userdata/UserDataManager.java]: why don't you just use the id (O(N log N) for the win!?

File: spilsdk/src/main/java/com/spilgames/spilsdk/userdata/UserDataManager.java [hunk 2]
```diff
@@ -0,0 +1,805 @@
...
+                        receivedPlayerCurrency.setDisplayDescription(currency.getDisplayDescription());
+                        receivedPlayerCurrency.setImageUrl(currency.getImageUrl());
+                        receivedPlayerCurrency.setInitialValue(currency.getInitialValue());
+                    }
+                }
+            }
+            remotePlayerData.put(Wallet, receivedWallet.BuildForJson().toJson(gson));
+
+            for (PlayerItem receivedPlayerItem : receivedInventory.getItemsMap().values()) {
+                for (Item item : spilGameData.getItemsMap().values()) {
```
Comment (svenvanhoenderdal) [inline spilsdk/src/main/java/com/spilgames/spilsdk/userdata/UserDataManager.java]: why don't you just use the id (O(N log N) for the win!?

File: spilsdk/src/main/java/com/spilgames/spilsdk/userdata/UserDataManager.java [hunk 3]
```diff
@@ -0,0 +1,805 @@
...
+            //Prepare and clean merge data for processing and sending
+            JSONObject mergeDataJSON = new JSONObject(mergeData);
+
+            JSONObject mergePlayerDataJSON = mergeDataJSON.getJSONObject(PlayerData);
+            JSONObject mergeGameStateJSON = mergeDataJSON.getJSONObject(GameState);
+
+            JSONObject mergeWallet = mergePlayerDataJSON.getJSONObject(Wallet);
+            JSONArray mergeCurrencies = new JSONArray();
+
+            for (int i = 0; i < mergeWallet.getJSONArray(PlayerDataManager.Currencies).length(); i++) {
```
Comment (svenvanhoenderdal) [inline spilsdk/src/main/java/com/spilgames/spilsdk/userdata/UserDataManager.java]: foreach FTW!

File: spilsdk/src/main/java/com/spilgames/spilsdk/userdata/UserDataManager.java [hunk 4]
```diff
@@ -0,0 +1,805 @@
...
+                currency.put(PlayerDataManager.Delta, mergeWallet.getJSONArray(PlayerDataManager.Currencies).getJSONObject(i).getInt(PlayerDataManager.Delta));
+                mergeCurrencies.put(currency);
+            }
+            mergeWallet.remove(PlayerDataManager.Currencies);
+            mergeWallet.put(PlayerDataManager.Currencies, mergeCurrencies);
+
+            JSONObject mergeInventory = mergePlayerDataJSON.getJSONObject(Inventory);
+            JSONArray mergeItems = new JSONArray();
+
+            for (int i = 0; i < mergeInventory.getJSONArray(PlayerDataManager.Items).length(); i++) {
```
Comment (svenvanhoenderdal) [inline spilsdk/src/main/java/com/spilgames/spilsdk/userdata/UserDataManager.java]: foreach FTW!

File: spilsdk/src/main/java/com/spilgames/spilsdk/userdata/UserDataManager.java [hunk 5]
```diff
@@ -0,0 +1,805 @@
...
+        if (userProfileString != null) {
+            UserData userData = PlayerDataInitialisation.initUserDataFromPrefs(context, playerDataManager.gson, userProfileString);
+
+            this.userData = userData;
+            return userData;
+        } else {
+            UserData userData = PlayerDataInitialisation.initUserDataFromAssets(context, playerDataManager.gson);
+
+            playerDataManager.storageUtil.putString(StorageUtil.Keys.SpilUserData, playerDataManager.gson.toJson(userData));
+            this.userData = userData;
```
Comment (svenvanhoenderdal) [inline spilsdk/src/main/java/com/spilgames/spilsdk/userdata/UserDataManager.java]: move lines out (code duplication with if in else)

File: spilsdk/src/main/java/com/spilgames/spilsdk/userdata/UserDataManager.java [hunk 6]
```diff
@@ -0,0 +1,805 @@
...
+     * Method used to update the data version of the current device.
+     *
+     * @param userData The local {@link UserData} information.
+     * @return The updated local {@link UserData} information.
+     */
+    public synchronized UserData updateUserDataVersion(UserData userData) {
+        String deviceId = UserIDGenerator.getUniqueDeviceId(context);
+
+        boolean check = false;
+        for (int i = 0; i < userData.getUserDataVersions().size(); i++) {
```
Comment (svenvanhoenderdal) [inline spilsdk/src/main/java/com/spilgames/spilsdk/userdata/UserDataManager.java]: foreach FTW!

**Comment by svenvanhoenderdal** [comment]
https://media0.giphy.com/media/gGWMc182QRkVa/giphy.gif

**Pull request merged**